### PR TITLE
FastBoot SSR deployment

### PR DESCRIPTION
**Pull request by jurgenwerk**
In the current state, docker container works, but need to test and configure the CI build step and waypoint deployment.

TODO: 
- [x] build infra for this (https://github.com/cardstack/infra/pull/87)
- [x] node-fetch in deployment/package.json version is outdated and should be upgraded
- [x] working action in the CI (need to test)
- [x] working waypoint "ssr-web" deployment (need to fill in the config, currently most of the config is copied from the hub app in waypoint)

### CODE REVIEW

File: .github/actions/deploy-ssr-web/action.yml [hunk 1]
```diff
@@ -0,0 +1,43 @@
+name: Deploy SSR web
+description: Deploy a FastBoot server serving a stripped down web-client
+
+inputs:
+  waypoint_version:
+    description: Waypoint version to use
+    required: false
+    default: '0.4.0'
+  environment:
```
Comment (jurgenwerk) [inline .github/actions/deploy-ssr-web/action.yml:9]: I think this is not needed anymore because we're always building a production build here

Comment (backspace) [reply .github/actions/deploy-ssr-web/action.yml:9]: In the web client deployment, it’s used for the [deployment script](https://github.com/cardstack/cardstack/blob/main/.github/actions/deploy-web-client/action.yml#L19) (which ultimately sets the `DEPLOY_TARGET` for Ember CLI Deploy) and [configuring Sentry](https://github.com/cardstack/cardstack/blob/main/.github/actions/deploy-web-client/action.yml#L41). Maybe we’ll need both of these? Not sure about how Hub URL is configured, will look more into it.

**Comment by jurgenwerk** [comment]
Making some progress, it fails because the VPC seems to be different - I think waypoint wants this to be deployed in the same VPC as hub-staging and waypoint-staging? 

``` 
» Deploying...
-> Creating new ECS cluster: ssr-web-staging
-> Created new ECS cluster: ssr-web-staging
-> Creating IAM role: ecr-ssr-web
-> Created IAM role: ecr-ssr-web
-> Found task IAM role to use: ssr-web-staging-ssr_web_ecr_task
-> Registering Task definition: waypoint-ssr-web
-> Registered Task definition: waypoint-ssr-web
-> Creating security group: ssr-web-inbound
-> Created security group: ssr-web-inbound
-> Authorizing ports to security group
-> Configured security group: ssr-web-inbound
-> Creating ALB target group
-> Created ALB target group
-> Using configured ALB Listener: arn:aws:elasticloadbalancing:us-east-1:680542703984:listener/app/ssr-web-staging/6ac46ddbf5b72dfa/ea02efa09592755f (load-balancer: arn:aws:elasticloadbalancing:us-east-1:680542703984:loadbalancer/app/ssr-web-staging/6ac46ddbf5b72dfa)
-> Modifying ALB Listener to introduce target group
! InvalidConfigurationRequest: The following target groups are in a different VPC
  than load balancer
  'arn:aws:elasticloadbalancing:us-east-1:680542703984:loadbalancer/app/ssr-web-staging/6ac46ddbf5b72dfa':
  arn:aws:elasticloadbalancing:us-east-1:680542703984:targetgroup/ssr-web-01FWE5EY8NYC320TTJ5FMEQM/ee73853a9882072d
  	status code: 400, request id: 9e[958](https://github.com/cardstack/cardstack/runs/5274289519?check_suite_focus=true#step:4:958)f81-9f98-4a03-9b76-117906294ede
```

**Comment by jurgenwerk** [comment]
Wow, the deploy command finally worked! However the https://ssr-wallet-staging.stack.cards/ is unreachable. Inspecting...

**Comment by jurgenwerk** [comment]
Progress update: 

1. The local deploy (`AWS_PROFILE=cardstack-staging waypoint up --app=ssr-web -plain`) and also CI deploy seems to work: https://ssr-wallet-staging.stack.cards/pay/sokol/123 opens OK 🍾

Next: 

1. Environment vars (cleanup in the actions file + do we need to insert any using `waypoint set`)? 
2. Move build commands to `push-main` (see https://github.com/cardstack/cardstack/pull/2698/files#diff-5dd9d290d705ab06a921e86eec40711d22a7b1fac32b8e4c477698fbe96d9f6aR49)

**Comment by backspace** [comment]
re updating `node-fetch`, I tried v3 but it [failed](https://github.com/cardstack/cardstack/runs/5294392942?check_suite_focus=true); the [documentation](https://github.com/node-fetch/node-fetch#commonjs) says:

> `node-fetch` from v3 is an ESM-only module - you are not able to import it with `require()`.

I wasn’t immediately sure how to address this, but maybe since there are already so many uses of 2.6.7 it’s fine to stick with it?

```
   - "workspace-aggregator-73681d76-700a-4dc2-9646-ff7c7226685c" depends on it
   - Hoisted from "_project_#ember-source-channel-url#node-fetch"
   - Hoisted from "_project_#discord.js#node-fetch"
   - Hoisted from "_project_#@cardstack#cardpay-cli#node-fetch"
   - Hoisted from "_project_#@cardstack#cardpay-subgraph#node-fetch"
   - Hoisted from "_project_#@cardstack#hub#node-fetch"
   - Hoisted from "_project_#ipfs-utils#node-fetch"
   - Hoisted from "_project_#@cardstack#cardpay-subgraph#@graphprotocol#graph-cli#node-fetch"
   - Hoisted from "_project_#web3-provider-engine#eth-json-rpc-infura#node-fetch"
   - Hoisted from "_project_#@cardstack#boxel#ember-fetch#node-fetch"
   - Hoisted from "_project_#web3-provider-engine#eth-json-rpc-middleware#node-fetch"
   - Hoisted from "_project_#firebase-walletconnect-push#firebase-tools#node-fetch"
   - Hoisted from "_project_#google-auth-library#gaxios#node-fetch"
   - Hoisted from "_project_#@cardstack#wc-provider#ethereum-test-network#hardhat#node-fetch"
   - Hoisted from "_project_#lerna#@lerna#version#@lerna#gitlab-client#node-fetch"
   - Hoisted from "_project_#@actions#github#@octokit#core#@octokit#request#node-fetch"
   - Hoisted from "_project_#firebase-admin#@google-cloud#firestore#google-gax#node-fetch"
   - Hoisted from "_project_#@cardstack#cardpay-subgraph#@graphprotocol#graph-cli#ipfs-http-client#ky-universal#node-fetch"
   - Hoisted from "_project_#firebase-walletconnect-push#firebase-tools#@google-cloud#pubsub#google-gax#node-fetch"
   - Hoisted from "_project_#firebase-admin#@google-cloud#storage#@google-cloud#common#teeny-request#node-fetch"
   - Hoisted from "_project_#@cardstack#hub#web3.storage#ipfs-car#ipfs-unixfs-importer#rabin-wasm#node-fetch"
```

**Comment by Aierie** [comment]
> re updating `node-fetch`, I tried v3 but it [failed](https://github.com/cardstack/cardstack/runs/5294392942?check_suite_focus=true); the [documentation](https://github.com/node-fetch/node-fetch#commonjs) says:
> 
> > `node-fetch` from v3 is an ESM-only module - you are not able to import it with `require()`.

Ya I think we should use v2 for now.

**Comment by jurgenwerk** [comment]
Progress: production deploys work ☑️ https://ssr-wallet.cardstack.com/pay/sokol/0xA458a7C13B33188b4F71aA47A1067f651284361E?amount=1&currency=CAD (deployed from my local)

**Review by backspace — APPROVED** [review]
I approve but also I am not an objective observer 😆

File: .github/workflows/push-main.yml [hunk 2]
```diff
@@ -238,6 +247,50 @@ jobs:
...
+            ${{ runner.os }}-yarn-
+      - name: Deploy ssr-web
+        uses: ./.github/actions/deploy-ssr-web
+        env:
+          HUB_URL: https://hub-staging.stack.cards
+          AWS_ACCESS_KEY_ID: ${{ secrets.STAGING_WAYPOINT_AWS_ACCESS_KEY }}
+          AWS_SECRET_ACCESS_KEY: ${{ secrets.STAGING_WAYPOINT_AWS_SECRET }}
+          AWS_DEFAULT_REGION: us-east-1
+          WAYPOINT_SERVER_TOKEN: ${{ secrets.STAGING_WAYPOINT_SERVER_TOKEN }}
+          WAYPOINT_SERVER_ADDR: ${{ secrets.STAGING_WAYPOINT_SERVER_ADDR }}
```
Comment (Aierie) [inline .github/workflows/push-main.yml:272]: Are these necessary?

File: packages/ssr-web/deployment/fastboot-server.js [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+  port: 4000, // Optional - Sets the port the server listens on (defaults to the PORT env var or 3000).
+  buildSandboxGlobals(defaultGlobals) {
+    // Optional - Make values available to the Ember app running in the FastBoot server, e.g. "MY_GLOBAL" will be available as "GLOBAL_VALUE"
+    return Object.assign({}, defaultGlobals, {
+      btoa: function (str) {
+        return Buffer.from(str).toString('base64');
+      },
+      fetch,
+    });
+  },
```
Comment (Aierie) [inline packages/ssr-web/deployment/fastboot-server.js:18]: In the future maybe we can find a way to share the sandbox global definition between this and the dev build?

**Review by Aierie — APPROVED** [review]

**Pull request merged**